BigInt (#258)
javascript-tutorial/it.javascript.info: 1-js/99-js-misc/05-bigint/article.md

[recent caniuse="bigint"]

> `BigInt` è uno speciale tipo numerico che supporta interi di lunghezza arbitraria.

Reviewer: ```suggestion
`BigInt` è uno speciale tipo numerico che supporta numeri interi di lunghezza arbitraria.
```

> Un bigint viene creato aggiungendo il suffisso `n` alla fine di un intero, oppure invocando la funzione `BigInt`, la quale crea bigints a partire da stringhe, numeri etc.

Reviewer: ```suggestion
Un bigint viene creato aggiungendo il suffisso `n` alla fine di un numero intero, oppure invocando la funzione `BigInt`, la quale crea bigints a partire da stringhe, numeri etc.
```

```js run
alert( 1 == 1n ); // true

alert( 1 === 1n ); // false
```

## Operazioni booleane

Quando utilizzati all'interno di un `if` o qualsiasi altra operazione booleana, i bigints si comportano come numbers.

Ad esempio, in `if`, bigint `0n` vale `false`, gli altri valori valgono `true`:

```js run
if (0n) {
  // non verrà mai eseguito
}
```

Operatori booleani, come `||`, `&&` e tutti gli altri, funzionano con i bigint in maniera simile ai number:

```js run
alert( 1n || 2 ); // 1 (1n viene considerato true)

alert( 0n || 2 ); // 2 (0n viene considerato false)
```

## Polyfills

Costruire un polyfill per bigints è difficile. Il motivo è che molti operatori JavaScript, come `+`, `-` e cosi via si comportano in maniera differente con i bigint rispetto ai numeri regolari.

Ad esempio, la divisione di bigint ritorna sempre un bigint (arrotondato se necessario).

Per poter emulare questo comportamento, un polyfill deve analizzare il codice e rimpiazzare tutti questi operatori con funzioni proprie. Fare questo può essere complesso e costerebbe molto in termini di performance.

Quindi, non esiste alcun polyfill ottimale.

> Anche se, un'alternativa è stata proposta dagli sviluppatori della libreria [JSBI](https://github.com/GoogleChromeLabs/jsbi).

Reviewer 2: ```suggestion
Comunque, un'alternativa è stata proposta dagli sviluppatori della libreria [JSBI](https://github.com/GoogleChromeLabs/jsbi).
```

> Questa libreria implementa i bigint utilizzando un proprio metodo. Possiamo utilizzare questi invece dei bigint integrati dal linguaggio:

Reviewer 2: ```suggestion
Questa libreria implementa i bigint utilizzando metodi propri. Possiamo utilizzare questi invece dei bigint integrati dal linguaggio:
```

| Operazione | `BigInt` integrati | JSBI |
|-----------|-----------------|------|
| Creazione da Number | `a = BigInt(789)` | `a = JSBI.BigInt(789)` |
| Addizione | `c = a + b` | `c = JSBI.add(a, b)` |
| Sottrazione	| `c = a - b` | `c = JSBI.subtract(a, b)` |
| ... | ... | ... |

> ...E poi utilizzare il polyfill (Babel plugin) per convertire le invocazioni a JSBI calls ai bigint integrati per i browser che li supportano.

Reviewer 2: ```suggestion
...quindi utilizzare il polyfill (Babel plugin) per convertire le invocazioni a JSBI in bigint nativi, per i browser che li supportano.
```

In altre parole, questo approccio suggerisce di scrivere il codice utilizzando JSBI piuttosto dei bigint integrati. JSBI funziona con i numbers proprio come i bigint integrati, emulandoli secondo quanto descritto nelle specifiche, quindi il codice sarà "bigint-ready".

> Possiamo utilizzare questo codice JSBI "cosi com'è" sia per i motori che non supportano i bigint che per quelli che li supportano - il polyfill convertirà le invocazioni in bigint integrati.

Reviewer: ```suggestion
Possiamo utilizzare questo codice JSBI "così com'è" sia per i motori che non supportano i bigint che per quelli che li supportano - il polyfill convertirà le invocazioni in bigint integrati.
```
